cannon: Finish emulating rest of 64-bit instructions

Inphi opened the pull request:
This completes 64-bit instruction execution (except lld/scd which is handled separately).

## Testing

- Differential tests for non-trivial 32-bit instructions asserting backwards compatibility.
- Added new tests for 64-bit instructions (except lld/scd).
- Fuzz tests for `dmult`

These tests are only run on by the 64-bit VM. So they do not run on CI, except for the `dmult` fuzz test which is executed by the `fuzz-golang` job.

## Meta

Fixes [#12243](https://github.com/ethereum-optimism/optimism/issues/12243) and [#12248](https://github.com/ethereum-optimism/optimism/issues/12248).

Inphi committed:
cannon: Finish emulating rest of 64-bit instructions

This fixes the 64-bit stubs for various instructions (except lld/scd).

mbaxter commented on go.mod:
```diff
@@ -47,11 +47,10 @@ require (
 	golang.org/x/sync v0.8.0
 	golang.org/x/term v0.25.0
 	golang.org/x/time v0.7.0
-	lukechampine.com/uint128 v1.3.0
```
There are a bunch of changes here - shouldn't the diff here just modify this one dependency?

Inphi replied:
It's a side effect of running `go mod tidy`. Sometimes other dependencies get affected even when removing deps, but this won't resolve in an incorrectly resolved dependency set.

mbaxter commented on cannon/mipsevm/exec/mips_instructions.go:
```diff
@@ -347,13 +347,24 @@ func ExecuteMipsInstruction(insn uint32, opcode uint32, fun uint32, rs, rt, mem
+			if arch.IsMips32 {
+				return mem
+			} else {
+				return SignExtend((mem>>(32-((rs&0x4)<<3)))&0xFFFFFFFF, 32)
```
Should we add a follow-up ticket to use some of the utils from [this PR](https://github.com/ethereum-optimism/optimism/pull/12419/files#diff-cbb25bcce1e91c91dbb8e0fc11d64d6c3fbf5ef89bf831a3a325734f962a7c39R573) to consolidate the load / store logic?

Inphi replied:
Indeed. Created an issue here - https://github.com/ethereum-optimism/optimism/issues/12562.

mbaxter commented on cannon/mipsevm/exec/mips_instructions.go:
```diff
@@ -378,18 +405,45 @@ func ExecuteMipsInstruction(insn uint32, opcode uint32, fun uint32, rs, rt, mem
+				return (mem & Word(^mask)) | val
+			} else {
+				// similar to the above mips32 implementation but constrained to the nearest 4-byte memory word
+				shift := 32 - (((rs >> 2) & 0x1) << 5)
```
(nit) Looks like there are 2 different styles for calculating the shift to find the 32-bit value:
- (val & 0x4) << 3
- ((val >> 2) & 0x1) << 5

Maybe we should standardize the approach - first one is probably more readable?

Inphi replied:
Aye. Done.

mbaxter commented on cannon/mipsevm/tests/evm_common64_test.go:
```diff
@@ -0,0 +1,539 @@
+		{name: "daddu. 32-bit. expect double-word sized x", funct: 0x2d, isImm: false, rs: Word(^uint32(0)), rt: Word(0x12), expectRes: Word(0x1_00_00_00_11)},        // dadu t0, s1, s2
+		{name: "daddu. doubleword-sized, word-sized", funct: 0x2d, isImm: false, rs: Word(0x0FFFFFFF_00000012), rt: Word(0x20), expectRes: Word(0x0FFFFFFF_00000032)}, // dadu t0, s1, s2
+		{name: "daddu. overflow", funct: 0x2d, isImm: false, rs: Word(12), rt: ^Word(0), expectRes: Word(11)},                                                         // dadu t0, s1, s2
+		{name: "daddu. overflow x", funct: 0x2d, isImm: false, rs: ^Word(0), rt: Word(12), expectRes: Word(11)},                                                       // dadu t0, s1, s2
```
what does the x mean here?

Inphi replied:
it means "crossed" operands. I'll give it a clearer name.

mbaxter commented on cannon/mipsevm/tests/evm_common64_test.go:
```diff
@@ -0,0 +1,539 @@
+			goVm := v.VMFactory(nil, os.Stdout, os.Stderr, testutil.CreateLogger(), testutil.WithRandomization(int64(i)), testutil.WithPC(0), testutil.WithNextPC(4))
+			state := goVm.GetState()
+			var insn uint32
+			var baseReg uint32 = 17
```
Should this be?:
```suggestion
			var rsReg uint32 = 17
```

Inphi replied:
Fixed

mbaxter commented on cannon/mipsevm/tests/evm_common64_test.go:
```diff
@@ -0,0 +1,539 @@
+		{name: "dsll32", funct: 0x3c, rd: Word(0xAA_BB_CC_DD_A1_B1_C1_D1), rt: Word(0xFF_FF_FF_FF_FF_FF_FF_FF), sa: 1, expectRes: Word(0xFF_FF_FF_FE_00_00_00_00)},  // dsll32 t8, s2, 1
+		{name: "dsll32", funct: 0x3c, rd: Word(0xAA_BB_CC_DD_A1_B1_C1_D1), rt: Word(0xFF_FF_FF_FF_FF_FF_FF_FF), sa: 31, expectRes: Word(0x80_00_00_00_00_00_00_00)}, // dsll32 t8, s2, 31
+
+		{name: "dsrl32", funct: 0x3e, rd: Word(0xAA_BB_CC_DD_A1_B1_C1_D1), rt: Word(0x1), sa: 0, expectRes: Word(0x0)},                                 // dsrl32 t8, s2, 0
+		{name: "dsrl32", funct: 0x3e, rd: Word(0xAA_BB_CC_DD_A1_B1_C1_D1), rt: Word(0x1), sa: 31, expectRes: Word(0x0)},                                // dsrl32 t8, s2, 31
```
Maybe worth adding a few test cases where `rt` is shifted left 32 bits:
```
{name: "dsrl32", funct: 0x3e, rd: Word(0xAA_BB_CC_DD_A1_B1_C1_D1), rt: Word(0x1_0000_0000), sa: 0, expectRes: Word(0x1)}, 
{name: "dsrl32", funct: 0x3e, rd: Word(0xAA_BB_CC_DD_A1_B1_C1_D1), rt: Word(0x1_0000_0000), sa: 31, expectRes: Word(0x0)},
```

Inphi replied:
added

mbaxter commented on cannon/mipsevm/tests/evm_common64_test.go:
```diff
@@ -0,0 +1,539 @@
+		{name: "dmult", funct: 0x1c, rs: 0x01_00_00_00_00, rt: 2, expectLo: 0x02_00_00_00_00, expectHi: 0},
+		{name: "dmult", funct: 0x1c, rs: 0x01_00_00_00_00_00_00_00, rt: 2, expectLo: 0x02_00_00_00_00_00_00_00, expectHi: 0},
+		{name: "dmult", funct: 0x1c, rs: 0x40_00_00_00_00_00_00_00, rt: 2, expectLo: 0x80_00_00_00_00_00_00_00, expectHi: 0x0},
+		{name: "dmult", funct: 0x1c, rs: 0x40_00_00_00_00_00_00_00, rt: 0x1000, expectLo: 0x0, expectHi: 0x4_00},
```
Probably worth adding some test cases that set both hi and lo.  For example:
```
{name: "dmult", funct: 0x1c, rs: 0x40_00_00_00_00_00_00_01, rt: 0x1000, expectLo: 0x1000, expectHi: 0x4_00},
```

Inphi replied:
added a couple more tests.

mbaxter commented on cannon/mipsevm/tests/evm_common64_test.go:
```diff
@@ -0,0 +1,539 @@
+		{name: "dmult", funct: 0x1c, rs: 0x01_00_00_00_00_00_00_00, rt: 2, expectLo: 0x02_00_00_00_00_00_00_00, expectHi: 0},
+		{name: "dmult", funct: 0x1c, rs: 0x40_00_00_00_00_00_00_00, rt: 2, expectLo: 0x80_00_00_00_00_00_00_00, expectHi: 0x0},
+		{name: "dmult", funct: 0x1c, rs: 0x40_00_00_00_00_00_00_00, rt: 0x1000, expectLo: 0x0, expectHi: 0x4_00},
+		{name: "dmult", funct: 0x1c, rs: 0x80_00_00_00_00_00_00_00, rt: 0x1000, expectLo: 0x0, expectHi: 0x8_00},
```
It's unclear to me from the spec whether `hi` should be sign extended here.  The fact that `mult` is sign-extended make me wonder if we should also sign-extend here.

Inphi replied:
Based on qemu-mips observations, it's conditional on the sign bits of the operands. The sign is extended if either, but not both, of the operands are. So it's basically regular arithmetic, with the sign extension used to preserve the negative result. Though as you'll see in this PR emulating this isn't as simple as doing the arithmetic in uint128. The two special cases for sign extension has to be explicitly handled.

mbaxter commented on cannon/mipsevm/tests/evm_common64_test.go:
```diff
@@ -0,0 +1,539 @@
+		{name: "dmultu", funct: 0x1d, rs: 0x80_00_00_00_00_00_00_00, rt: 0x80_00_00_00_00_00_00_00, expectLo: 0x0, expectHi: 0x40_00_00_00_00_00_00_00},
+
+		// ddiv rs, rt
+		{name: "ddiv", funct: 0x1e, rs: 0, rt: 1, expectLo: 0, expectHi: 0},
```
The vm panics on divide by zero - is that what we want?  Probably worth adding a test around that to make sure the on- and off-chain impls are consistent.

Inphi replied:
Good point. I've added a todo to do this for the solidity VM.

mbaxter commented on cannon/mipsevm/tests/evm_common64_test.go:
```diff
@@ -0,0 +1,539 @@
+
+	v := GetMultiThreadedTestCase(t)
+	for i, tt := range cases {
+		testName := fmt.Sprintf("%v %v", v.Name, tt.name)
+		t.Run(testName, func(t *testing.T) {
```
(nit) Since we're only running these for multi-threaded, you can just use the plain test name throughout:
```suggestion
		t.Run(tt.name, func(t *testing.T) {
```

Inphi replied:
I'll leave it as-is so it's super explicit what vm is being tested here.

mbaxter commented on cannon/mipsevm/tests/fuzz_evm_common64_test.go:
```diff
@@ -0,0 +1,54 @@
+	f.Fuzz(func(t *testing.T, rs arch.Word, rt arch.Word, seed int64) {
+		for _, v := range versions {
+			t.Run(v.Name, func(t *testing.T) {
+				goVm := v.VMFactory(nil, os.Stdout, os.Stderr, testutil.CreateLogger(), testutil.WithRandomization(seed), testutil.WithPC(0), testutil.WithNextPC(4))
```
just fyi - we also have this `WithPCAndNextPC` util for less noise:
```suggestion
				goVm := v.VMFactory(nil, os.Stdout, os.Stderr, testutil.CreateLogger(), testutil.WithRandomization(seed), testutil.WithPCAndNextPC(0))
```

mbaxter commented on cannon/mipsevm/tests/fuzz_evm_common64_test.go:
```diff
@@ -0,0 +1,54 @@
+
+func FuzzStateDmultInsn(f *testing.F) {
+	versions := GetMipsVersionTestCases(f)
+	f.Fuzz(func(t *testing.T, rs arch.Word, rt arch.Word, seed int64) {
```
If I change the param type to a signed int, I start to get test failure here:
```suggestion
	f.Fuzz(func(t *testing.T, rs int64, rt int64, seed int64) {
```

When running this locally with the existing `Word` (unsigned) operands, the fuzzer seems to only be generating 1 byte values between 0 and 255.

mbaxter commented on cannon/mipsevm/tests/evm_common64_test.go:
```diff
@@ -0,0 +1,539 @@
+//go:build cannon64
+// +build cannon64
+
+package tests
```
awesome test coverage 🎉

mbaxter reviewed (commented):
Looking good!  Left a number of nits + a few questions.

Inphi committed:
review comments; fix dmult

Inphi committed:
add todo

Inphi commented:
@mbaxter I fixed a bug in `dmult`. And I've removed the `dmult` fuzz tests in favor of more static tests. I couldn't figure out a way to reimplement dmult correctly using the Go std library. At least not without copying the implementation used by mipsevm, which defeats the purpose of the fuzz test.

I have verified the expected `dmult` behavior using user-mode qemu like I mentioned earlier. All static test cases were written based off the results of qemu-mips. Please let me know if there are any more interesting cases you'd like me to add.

Inphi committed:
test div by zero

Inphi committed:
cannon: Fix remaining mips64 emulation bugs

mbaxter commented on cannon/mipsevm/exec/mips_instructions.go:
```diff
@@ -571,7 +572,33 @@ func HandleHiLo(cpu *mipsevm.CpuScalars, registers *[32]Word, fun uint32, rs Wor
+			absB = 0x8000000000000000
+		} else {
+			absB = uint64(abs64(b))
+		}
```
Looks like this special-casing is unnecessary.  Tests are passing if I skip this logic.

Inphi replied:
Yeah you're right. `-0x8000000000000000` requires more than 64-bits to represent so this is dead code. Removed special case.

Inphi replied:
And if you extend -0x8000000000000000 to u128, it's still 0x8000000000000000 with the sign bit extended to 128-bits. So the math still checks out for the absolute value here.

mbaxter commented on cannon/mipsevm/exec/mips_instructions.go:
```diff
@@ -581,10 +608,16 @@ func HandleHiLo(cpu *mipsevm.CpuScalars, registers *[32]Word, fun uint32, rs Wor
 		cpu.LO = Word(lo)
 	case 0x1e: // ddiv
 		assertMips64Fun(fun)
+		if rt == 0 {
+			panic("instruction divide by zero")
+		}
```
Should we also go ahead and add this constraint + tests for div/divu?

Inphi replied:
Yes. I just realized I only added this for the double-word div instructions.

Inphi replied:
div/divu 32-bit tests were recently implemented by https://github.com/ethereum-optimism/optimism/pull/12459.
To avoid rebasing other branches that depend on this PR, I'll fix this in a separate PR.

Inphi replied:
Opened a PR to do this here - https://github.com/ethereum-optimism/optimism/pull/12600

mbaxter commented on cannon/mipsevm/tests/evm_common64_test.go:
```diff
@@ -459,34 +461,49 @@ func TestEVMSingleStep_LoadStore64(t *testing.T) {
+		{name: "dmult 6", funct: 0x1c, rs: 0x80_00_00_00_00_00_00_00, rt: 0x1000, expectLo: 0x0, expectHi: 0xFF_FF_FF_FF_FF_FF_F8_00},
+		{name: "dmult 7", funct: 0x1c, rs: 0x80_00_00_00_00_00_00_00, rt: 0x80_00_00_00_00_00_00_00, expectLo: 0x0, expectHi: 0x40_00_00_00_00_00_00_00},
+		{name: "dmult 8", funct: 0x1c, rs: 0x40_00_00_00_00_00_00_01, rt: 0x1000, expectLo: 0x1000, expectHi: 0x4_00},
+		{name: "dmult 9", funct: 0x1c, rs: 0x80_00_00_00_00_00_00_80, rt: 0x80_00_00_00_00_00_00_80, expectLo: 0x4000, expectHi: 0x3F_FF_FF_FF_FF_FF_FF_80},
```
This one was hard for me to parse so double-checked with "big/math" and I'm getting a different result here for some reason: `0x4000_0000_0000_0080_0000_0000_0000_4000`:
```
func main() {
	val := "8000000000000080"
	bigmult(val, val)
}

func bigmult(aVal, bVal string) {
    a := new(big.Int)
    b := new(big.Int)

    // Set the values of a and b using strings
    _, ok := a.SetString(aVal, 16)
    if !ok {
        fmt.Println("Error parsing number a")
        return
    }
    _, ok = b.SetString(bVal, 16)
    if !ok {
        fmt.Println("Error parsing number b")
        return
    }

    fmt.Printf("0x%v x 0x%v\n", a.Text(16), b.Text(16))
    result := new(big.Int)
    result.Mul(a, b)

    // Print the result
    fmt.Printf("Result: 0x%v", result.Text(16))
}
```

Inphi replied:
big/math doesn't treat the 0x80_00_00_00_00_00_00_80 as negative. Instead, it performs the multiply as if the operands were 128-bit integers. Which is not what we want.

Here's how you'd implement this in Rust using 128-bit intrinsics -https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5f381fa079109d5a6e469917e298c7e7.

mbaxter commented on cannon/mipsevm/tests/evm_common64_test.go:
```diff
@@ -459,34 +461,49 @@ func TestEVMSingleStep_LoadStore64(t *testing.T) {
+		{name: "dmult 8", funct: 0x1c, rs: 0x40_00_00_00_00_00_00_01, rt: 0x1000, expectLo: 0x1000, expectHi: 0x4_00},
+		{name: "dmult 9", funct: 0x1c, rs: 0x80_00_00_00_00_00_00_80, rt: 0x80_00_00_00_00_00_00_80, expectLo: 0x4000, expectHi: 0x3F_FF_FF_FF_FF_FF_FF_80},
+		{name: "dmult 10", funct: 0x1c, rs: Word(0xFF_FF_FF_FF_FF_FF_FF_FF), rt: Word(0x1), expectLo: 0xFF_FF_FF_FF_FF_FF_FF_FF, expectHi: 0xFF_FF_FF_FF_FF_FF_FF_FF},
+		{name: "dmult 11", funct: 0x1c, rs: Word(0xFF_FF_FF_FF_FF_FF_FF_FF), rt: Word(0xFF_FF_FF_FF_FF_FF_FF_FF), expectLo: 0x1, expectHi: Word(0)},
+		{name: "dmult 12", funct: 0x1c, rs: Word(0xFF_FF_FF_FF_FF_FF_FF_D3), rt: Word(0xAA_BB_CC_DD_A1_D1_C1_E0), expectLo: 0xFC_FC_FD_0A_8E_20_EB_A0, expectHi: 0x00_00_00_00_00_00_00_0E},
+		{name: "dmult 13", funct: 0x1c, rs: Word(0x7F_FF_FF_FF_FF_FF_FF_FF), rt: Word(0xAA_BB_CC_DD_A1_D1_C1_E1), expectLo: 0xD5_44_33_22_5E_2E_3E_1F, expectHi: 0xD5_5D_E6_6E_D0_E8_E0_F0},
+		{name: "dmult 14", funct: 0x1c, rs: Word(0x7F_FF_FF_FF_FF_FF_FF_FF), rt: Word(0x8F_FF_FF_FF_FF_FF_FF_FF), expectLo: 0xF0_00_00_00_00_00_00_01, expectHi: 0xC7_FF_FF_FF_FF_FF_FF_FF},
```
maybe worth repeating all these test cases for dmult?

mbaxter reviewed (approved):
lgtm - just a few last questions on dmult and div/divu divide by zero

Inphi committed:
add a couple more dmultu tests

Inphi committed:
remove dead code

Inphi committed:
fix 64-bit Makefile build script; review comments

Inphi committed:
fix build script

Inphi committed:
Merge pull request #12569 from ethereum-optimism/inphi/cannon64-test-programs

cannon: Fix remaining mips64 emulation bugs
